### DISCUSSION TITLE
Solved: Debugging a possible upgrade issue from 0.4.1 to 0.5.0

### DISCUSSION THREAD

**Discussion by Taitava** [Taitava/obsidian-shellcommands#63] [Debugging and testing]
This discussion started originally in https://github.com/Taitava/obsidian-shellcommands/discussions/16 . As the discussion evolved quite long on discussing about a possible upgrade bug, I decided to start a new discussion, because I'm writing quite a long post about testing this possible bug.

> After restarting Obsidian (close and open) the fields are `ignore` !!!!!
> 
>     * I was going to start selecting ignore to prepare the data.json and all shell commands are `ignore` again.
> 
> 
> What I did too before writing this to you @Taitava:
> 
>     * I changed one shell command to `Current File: caret position`
> 
>     * Restart Obsidian to see if it will save my change
> 
> 
> Perfect behavior: ignore` for all shell commands except the one that I changed.

Hmm, so the output settings change to *Ignore* for all commands (except the one you mentioned) without you needing to change them manually. It's odd. Even if this seems to be ok now, I need to test this more in the next few days. While a restart might solve some issues, it should not be needed. The plugin _should_ work after an upgrade without needing Obsidian to be restarted.

I'm going to test this using the following method, and you can test this too, if you want and have time. The idea is just to see can we get the same problem occur again. If we can reproduce it, then it can possibly be fixed.

1. Close Obsidian.
2. Go to folder `.obsidian/plugins/obsidian-shellcommands/`.
3. Make a backup copy of `data.json`, name the copy e.g. `data-original.json`. Without this copy, your commands might be lost if some errors happen during this test.
4. Open `data.json` in a text editor. It's probably quite long when you have many commands in it, but if you read it with time, you'll understand approximately how the file's structure works for defining each shell command.
5. Let's take a look at this small example `data.json`:

```json
{
  "working_directory": "",
  "preview_variables_in_command_palette": true,
  "shell_commands": {
    "0": {
      "shell_command": "echo \"Hello world!\"",
      "alias": "",
      "confirm_execution": false,
      "ignore_error_codes": [],
      "output_channels": {
        "stdout": "ignore",
        "stderr": "notification"
      },
      "output_channel_order": "stdout-first"
    },
    "1": {
      "shell_command": "echo \"Another command\"",
      "alias": "",
      "confirm_execution": false,
      "ignore_error_codes": [],
      "output_channels": {
        "stdout": "ignore",
        "stderr": "notification"
      },
      "output_channel_order": "stdout-first"
    }
  },
  "error_message_duration": 20,
  "notification_message_duration": 10,
  "commands": []
}
```

The lines that start with a number in quotes (`"0": {` and `"1": {` in this example) are start lines for a shell command. For example, the following part defines a single shell command:
```json
    "0": {
      "shell_command": "echo \"Hello world!\"",
      "alias": "",
      "confirm_execution": false,
      "ignore_error_codes": [],
      "output_channels": {
        "stdout": "ignore",
        "stderr": "notification"
      },
      "output_channel_order": "stdout-first"
    },
```

Now, for each shell command, you can do these two things:
1. Remove this small part:
    ```json
          "output_channels": {
        "stdout": "ignore",
        "stderr": "notification"
      },
      "output_channel_order": "stdout-first"
    ```
2. On the line `"ignore_error_codes": [],` remove the trailing comma `,` so that the line becomes: `"ignore_error_codes": []`

Example result for one command:
```json
    "0": {
      "shell_command": "echo \"Hello world!\"",
      "alias": "",
      "confirm_execution": false,
      "ignore_error_codes": []
    },
```

Example result for a complete file:
```json
{
  "working_directory": "",
  "preview_variables_in_command_palette": true,
  "shell_commands": {
    "0": {
      "shell_command": "echo \"Hello world!\"",
      "alias": "",
      "confirm_execution": false,
      "ignore_error_codes": []
    },
    "1": {
      "shell_command": "echo \"Another command\"",
      "alias": "",
      "confirm_execution": false,
      "ignore_error_codes": []
    }
  },
  "error_message_duration": 20,
  "notification_message_duration": 10,
  "commands": []
}
```

When you have done this, the `data.json` file should have a quite similar structure as it had when SC version was `0.4.1`. Now you can create a second copy of `data.json`, name it e.g. `data-modified.json` or something. When you open Obsidian next time, SC will automatically add all the fields to `data.json` that you just removed. The test idea is, that now you can see does SC set output **again** to *Current file, caret position*, or will it correctly use *Ignore* this time. Now that you have `data-modified.json`, you can repeat the test (if needed) to check again how the new fields' addition goes, just by copying `data-modified.json` over `data.json` and starting Obsidian.

SC always adds missing fields to the config file, so you can remove pretty much whatever field in the file and SC will recreate the field with a default value. But **beware, all the special characters like `{` `}` `[` `]` `:` `,` `"` need to be excatly correct**. (That's why I mentioned that the last comma `,` should be deleted). If there is a small error with these characters, Obsidian will not load the settings file at all, and it does not even tell that there was a problem. If you notice all your commands disappearing in the settings panel (or in command palette), close Obsidian and copy/rename `data-original.json` back to `data.json` and start Obsidian again.

**Comment by FelipeRearden** [comment]
Hello @Taitava !!!!

Now I can confirm with 100% of certainty that everytinhg is OK with `0.5.0`.
- I have a Windows machine that I use to test CSS snippets
- I remember that I have Shell ommands there from my tests since your first version
- It was version `0.4.0` !!!
- I save the `data.json` from `0.4.0` version and update to `0.5.0` 

I can confirm that the default option to all Shell Commands (old and new) are `ignore` 

I restart Obsdian and again `ignore`

I changed some commands to `current position` and `notification` and it keep the value after restart Obsidian.

I have the `data.json` from both versions if you need :)


Sorry again for the false alarm @Taitava . I will pay more attention in the next versions :)

I wish you a fantastic Monday :)

**Reply by Taitava** [reply]
No problem! 🙂 It's good to ask if you are uncertain! 😉

**Reply by FelipeRearden** [reply]
Hello @Taitava !!!!

Everything is OK with this topic when updating `0.5.0` --> `0.5.1`.

Have a great day!

**Reply by Taitava** [reply]
Yea, so I expected, as `0.5.1` did not touch the settings. 🙂

**Reply by FelipeRearden** [reply]
Hello @Taitava !!!!

Everything is OK with this topic when updating `0.5.1` --> `0.6.0` on MacOS and Obsidian 0.12.15

What a great update 🎉 🎉   !!!!!!!!!!!!

**Reply by Taitava** [reply]
Thank you! 🌞
